File: backend/src/main/java/com/joying/chat/service/WebPushService.java
package com.joying.chat.service;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.joying.chat.domain.PushSubscription;
import com.joying.chat.dto.PushSubscriptionRequest;
import com.joying.chat.repository.PushSubscriptionRepository;

import nl.martijndwars.webpush.Notification;
import nl.martijndwars.webpush.PushService;

/**
 * 브라우저 푸시 알림.
 *
 * <p>VAPID 키가 없으면 {@code pushService}가 없다. 그럴 때는 보내지 않고 조용히
 * 넘어간다. 알림이 안 가는 것보다 그것 때문에 대화가 막히는 쪽이 나쁘다.
 */
@Service
public class WebPushService {

	private static final Logger log = LoggerFactory.getLogger(WebPushService.class);

	/** 한 사람이 여러 기기에서 구독할 수 있지만 무한정 쌓이게 두지 않는다 */
	private static final int MAX_SUBSCRIPTIONS_PER_USER = 10;

	private final PushSubscriptionRepository pushSubscriptionRepository;
	private final PushService pushService;
	private final ObjectMapper objectMapper;

	public WebPushService(PushSubscriptionRepository pushSubscriptionRepository,
						  @Autowired(required = false) PushService pushService,
						  ObjectMapper objectMapper) {
		this.pushSubscriptionRepository = pushSubscriptionRepository;
		this.pushService = pushService;
		this.objectMapper = objectMapper;
	}

	public boolean isPushEnabled() {
		return pushService != null;
	}

	/**
	 * 구독을 등록한다.
	 *
	 * <p>같은 주소로 다시 오면 새로 만들지 않고 갱신한다. 브라우저가 키를 바꿔 다시
	 * 구독하는 일이 있어서, 새로 만들면 죽은 구독이 쌓인다.
	 */
	@Transactional
	public PushSubscription subscribe(Long memberId, PushSubscriptionRequest request) {
		Optional<PushSubscription> existing =
			pushSubscriptionRepository.findByEndpoint(request.getEndpoint());

		if (existing.isPresent()) {
			PushSubscription subscription = existing.get();
			subscription.setMemberId(memberId);
			subscription.setP256dh(request.getP256dh());
			subscription.setAuth(request.getAuth());
			subscription.setUserAgent(request.getUserAgent());
			log.info("푸시 구독 갱신: memberId={}", memberId);
			return pushSubscriptionRepository.save(subscription);
		}

		List<PushSubscription> existingSubscriptions =
			pushSubscriptionRepository.findByMemberId(memberId);

		if (existingSubscriptions.size() >= MAX_SUBSCRIPTIONS_PER_USER) {
			existingSubscriptions.stream()
				.min(Comparator.comparing(PushSubscription::getCreatedAt))
				.ifPresent(oldest -> {
					log.warn("구독 개수 제한 초과로 가장 오래된 것을 지운다: memberId={}, createdAt={}",
						memberId, oldest.getCreatedAt());
					pushSubscriptionRepository.delete(oldest);
				});
		}

		PushSubscription subscription = new PushSubscription(
			memberId, request.getEndpoint(), request.getP256dh(),
			request.getAuth(), request.getUserAgent());

		log.info("새 푸시 구독 생성: memberId={}, 총 구독 개수={}",
			memberId, existingSubscriptions.size() + 1);
		return pushSubscriptionRepository.save(subscription);
	}

	@Transactional
	public void unsubscribe(String endpoint) {
		pushSubscriptionRepository.deleteByEndpoint(endpoint);
	}

	public List<PushSubscription> getSubscriptions(Long memberId) {
		return pushSubscriptionRepository.findByMemberId(memberId);
	}

	/**
	 * 알림을 보낸다.
	 *
	 * <p>보내는 동안 부르는 쪽을 잡아 두지 않는다. 알림이 늦는 것과 메시지 저장이
	 * 늦는 것은 값이 다르다.
	 */
	@Async
	public void sendNotification(Long memberId, PushNotificationPayload payload) {
		if (!isPushEnabled()) {
			return;
		}

		try {
			List<PushSubscription> subscriptions =
				pushSubscriptionRepository.findByMemberId(memberId);
			if (subscriptions.isEmpty()) {
				return;
			}

			String payloadJson = objectMapper.writeValueAsString(payload);
			for (PushSubscription subscription : subscriptions) {
				sendTo(memberId, subscription, payloadJson);
			}
		} catch (Exception e) {
			log.error("푸시 알림 전송 실패: memberId={}", memberId, e);
		}
	}

	/**
	 * 한 구독에 보낸다.
	 *
	 * <p>브라우저가 구독을 버렸으면 상대가 그것을 상태 코드로 알려 준다. 그때 우리
	 * 기록도 같이 지운다. 안 지우면 죽은 구독에 계속 보내게 된다.
	 */
	private void sendTo(Long memberId, PushSubscription subscription, String payloadJson) {
		try {
			Notification notification = new Notification(
				subscription.getEndpoint(), subscription.getP256dh(),
				subscription.getAuth(), payloadJson);

			int statusCode = pushService.send(notification).getStatusLine().getStatusCode();

			if (statusCode == 410 || statusCode == 404 || statusCode == 400) {
				log.warn("유효하지 않은 구독이라 지운다: memberId={}, statusCode={}", memberId, statusCode);
				pushSubscriptionRepository.delete(subscription);
			} else if (statusCode < 200 || statusCode >= 300) {
				log.error("푸시 알림 전송 실패: memberId={}, statusCode={}", memberId, statusCode);
			}
		} catch (Exception e) {
			log.error("푸시 알림 전송 중 오류: memberId={}", memberId, e);
		}
	}
}

File: backend/src/main/kotlin/com/joying/chat/service/WebPushService.kt
package com.joying.chat.service

import com.fasterxml.jackson.databind.ObjectMapper
import com.joying.chat.domain.PushSubscription
import com.joying.chat.dto.PushSubscriptionRequest
import com.joying.chat.repository.PushSubscriptionRepository
import nl.martijndwars.webpush.Notification
import nl.martijndwars.webpush.PushService
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * Web Push 알림 서비스
 */
@Service
class WebPushService(
    private val pushSubscriptionRepository: PushSubscriptionRepository,
    private val pushService: PushService?,
    private val objectMapper: ObjectMapper
) {
    private val logger = LoggerFactory.getLogger(javaClass)

    /**
     * 푸시 알림 서비스 활성화 여부 확인
     */
    fun isPushEnabled(): Boolean = pushService != null

    /**
     * 푸시 구독 등록
     *
     * 동일한 사용자가 여러 기기/브라우저에서 구독할 수 있으므로,
     * endpoint 기준으로 갱신하되, VAPID 키 변경에 대응하기 위해
     * 오래된 구독은 자동으로 정리됩니다.
     */
    @Transactional
    fun subscribe(memberId: Long, request: PushSubscriptionRequest): PushSubscription {
        // 1. 동일한 endpoint가 있으면 업데이트 (재구독 케이스)
        val existingByEndpoint = pushSubscriptionRepository.findByEndpoint(request.endpoint)
        if (existingByEndpoint.isPresent) {
            val subscription = existingByEndpoint.get()
            subscription.memberId = memberId
            subscription.p256dh = request.p256dh
            subscription.auth = request.auth
            subscription.userAgent = request.userAgent
            logger.info("푸시 구독 갱신: memberId=$memberId, endpoint=${request.endpoint.take(50)}...")
            return pushSubscriptionRepository.save(subscription)
        }

        // 2. 동일 사용자의 구독 개수 확인 (너무 많으면 오래된 것 정리)
        val existingSubscriptions = pushSubscriptionRepository.findByMemberId(memberId)
        val maxSubscriptionsPerUser = 10 // 사용자당 최대 10개 기기/브라우저

        if (existingSubscriptions.size >= maxSubscriptionsPerUser) {
            // 가장 오래된 구독 삭제 (createdAt 기준)
            val oldestSubscription = existingSubscriptions.minByOrNull { it.createdAt }
            if (oldestSubscription != null) {
                logger.warn("구독 개수 제한 초과 - 가장 오래된 구독 삭제: memberId=$memberId, endpoint=${oldestSubscription.endpoint.take(50)}..., createdAt=${oldestSubscription.createdAt}")
                pushSubscriptionRepository.delete(oldestSubscription)
            }
        }

        // 3. 새로운 구독 생성
        val subscription = PushSubscription(
            memberId = memberId,
            endpoint = request.endpoint,
            p256dh = request.p256dh,
            auth = request.auth,
            userAgent = request.userAgent
        )
        logger.info("새 푸시 구독 생성: memberId=$memberId, endpoint=${request.endpoint.take(50)}..., 총 구독 개수=${existingSubscriptions.size + 1}")
        return pushSubscriptionRepository.save(subscription)
    }

    /**
     * 푸시 구독 해제
     */
    @Transactional
    fun unsubscribe(endpoint: String) {
        pushSubscriptionRepository.deleteByEndpoint(endpoint)
    }

    /**
     * 회원의 모든 구독 조회
     */
    fun getSubscriptions(memberId: Long): List<PushSubscription> {
        return pushSubscriptionRepository.findByMemberId(memberId)
    }

    /**
     * 특정 회원에게 푸시 알림 전송 (비동기)
     */
    @Async
    fun sendNotification(memberId: Long, payload: PushNotificationPayload) {
        // 푸시 서비스가 비활성화되어 있으면 건너뛰기
        if (!isPushEnabled()) {
            logger.debug("푸시 알림 서비스가 비활성화되어 있습니다 (VAPID 키 미설정)")
            return
        }

        try {
            val subscriptions = pushSubscriptionRepository.findByMemberId(memberId)
            if (subscriptions.isEmpty()) {
                logger.info("푸시 알림 건너뜀 (구독 정보 없음): memberId=$memberId, title=${payload.title}")
                return
            }

            logger.info("푸시 알림 전송 시작: memberId=$memberId, 구독 개수=${subscriptions.size}, title=${payload.title}")

            val payloadJson = objectMapper.writeValueAsString(payload)

            subscriptions.forEach { subscription ->
                try {
                    val notification = Notification(
                        subscription.endpoint,
                        subscription.p256dh,
                        subscription.auth,
                        payloadJson
                    )
                    val response = pushService!!.send(notification)

                    // 410 Gone: 구독이 만료되었거나 삭제됨
                    // 404/400: 잘못된 구독 정보
                    // 이러한 상태 코드가 반환되면 구독이 더 이상 유효하지 않으므로 삭제
                    if (response.statusLine.statusCode in listOf(410, 404, 400)) {
                        logger.warn("푸시 알림 실패 - 유효하지 않은 구독 (${response.statusLine.statusCode}): memberId=$memberId, endpoint=${subscription.endpoint.take(50)}..., 구독 정보 삭제")
                        pushSubscriptionRepository.delete(subscription)
                    }
                    // 성공
                    else if (response.statusLine.statusCode in 200..299) {
                        logger.info("푸시 알림 전송 성공: memberId=$memberId, statusCode=${response.statusLine.statusCode}, endpoint=${subscription.endpoint.take(50)}...")
                    }
                    // 기타 오류
                    else {
                        logger.error("푸시 알림 전송 실패 - 알 수 없는 오류: statusCode=${response.statusLine.statusCode}, memberId=$memberId, endpoint=${subscription.endpoint}")
                    }
                } catch (e: Exception) {
                    logger.error("푸시 알림 전송 중 오류: memberId=$memberId, endpoint=${subscription.endpoint}", e)
                }
            }
        } catch (e: Exception) {
            logger.error("푸시 알림 전송 실패: memberId=$memberId", e)
        }
    }
}

/**
 * 푸시 알림 페이로드
 */
data class PushNotificationPayload(
    val title: String,
    val body: String,
    val icon: String? = null,
    val image: String? = null,
    val badge: String? = null,
    val tag: String? = null,
    val data: Map<String, Any> = emptyMap()
)